File: cpp/lc111.cpp
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

#include <deque>
using namespace std;
class Solution {
public:
    int minDepth(TreeNode* root) {
        deque<TreeNode*> deq;
        int depth = 0;
        if (root) deq.push_back(root);
        while (!deq.empty()) {
            depth += 1;
            int len = deq.size();
            for (int i = 0; i < len; i++) {
                TreeNode* node = deq.front();
                deq.pop_front();
                if (!node->right && !node->left) return depth;
                if (node->right) deq.push_back(node->right);
                if (node->left) deq.push_back(node->left);
            }
        }

        return depth;
    }
};

File: src/bin/lc111.rs
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
  pub val: i32,
  pub left: Option<Rc<RefCell<TreeNode>>>,
  pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
  #[inline]
  pub fn new(val: i32) -> Self {
    TreeNode {
      val,
      left: None,
      right: None
    }
  }
}

struct Solution;
fn main() {}
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::VecDeque;

impl Solution {
    pub fn min_depth(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let mut deque = VecDeque::new();
        let mut depth = 0;
        if let Some(node) = root {
            deque.push_back(node);
        }

        while !deque.is_empty() {
            let len = deque.len();
            depth += 1;
            for _ in 0..len {
                if let Some(node) = deque.pop_front() {
                    if node.borrow().right.is_none() && node.borrow().left.is_none() {
                        return depth;
                    }
                    if let Some(left) = node.clone().borrow_mut().left.take() {
                        deque.push_back(left);
                    }
                    if let Some(right) = node.clone().borrow_mut().right.take() {
                        deque.push_back(right);
                    }
                }
            }
        }
        depth
    }
}